metricsx/middleware.go

```diff
@@ -71,256 +44,90 @@ func Hash(value string) string {
-			l.Warn(`Environment variable "DISABLE_TELEMETRY" has been deprecated and will be removed in an upcoming release. Use configuration key "sqa.opt_out: true" or environment variable "SQA_OPT_OUT=true" instead.`)
-		}
-	}
+func anonymizePath(whitelist []string, path string, salt string) string {
+	path = strings.ToLower(path)
 
-	if !optOut {
-		l.Info("Software quality assurance features are enabled. Learn more at: https://www.ory.sh/docs/ecosystem/sqa")
-		oi = analytics.OSInfo{
-			Version: fmt.Sprintf("%s-%s", runtime.GOOS, runtime.GOARCH),
+	for _, p := range whitelist {
+		p = strings.ToLower(p)
+		if len(path) == len(p) && path[:len(p)] == strings.ToLower(p) {
+			return p
+		} else if len(path) > len(p) && path[:len(p)+1] == strings.ToLower(p)+"/" {
+			return path[:len(p)] + "/" + Hash(path[len(p):]+"|"+salt)
 		}
 	}
 
-	m := &Service{
-		optOut: optOut,
-		salt:   uuid.New(),
-		o:      o,
-		c:      segment,
-		l:      l,
-		mem:    new(MemoryStatistics),
-		context: &analytics.Context{
-			IP: net.IPv4(0, 0, 0, 0),
-			App: analytics.AppInfo{
-				Name:    o.Service,
-				Version: o.BuildVersion,
-				Build:   fmt.Sprintf("%s/%s/%s", o.BuildVersion, o.BuildHash, o.BuildTime),
-			},
-			OS: oi,
-			Traits: analytics.NewTraits().
-				Set("optedOut", optOut).
-				Set("instanceId", uuid.New()).
-				Set("isDevelopment", o.IsDevelopment),
-			UserAgent: "github.com/ory/x/metricsx.Service/v0.0.1",
-		},
-	}
-
-	instance = m
-
-	go m.Identify()
-	go m.ObserveMemory()
-
-	return m
-}
-
-// Identify enables reporting to segment.
-func (sw *Service) Identify() {
-	if err := resilience.Retry(sw.l, time.Minute*5, time.Hour*24*30, func() error {
-		return sw.c.Enqueue(analytics.Identify{
-			UserId:  sw.o.ClusterID,
-			Traits:  sw.context.Traits,
-			Context: sw.context,
-		})
-	}); err != nil {
-		sw.l.WithError(err).Debug("Could not commit anonymized environment information")
-	}
+	return "/"
 }
 
-// ObserveMemory commits memory statistics to segment.
-func (sw *Service) ObserveMemory() {
-	if sw.optOut {
-		return
-	}
-
-	for {
-		sw.mem.Update()
-		if err := sw.c.Enqueue(analytics.Track{
-			UserId:     sw.o.ClusterID,
-			Event:      "memstats",
-			Properties: analytics.Properties(sw.mem.ToMap()),
-			Context:    sw.context,
-		}); err != nil {
-			sw.l.WithError(err).Debug("Could not commit anonymized telemetry data")
+func anonymizeQuery(query url.Values, salt string) string {
+	for _, q := range query {
+		for i, s := range q {
+			if s != "" {
+				s = Hash(s + "|" + salt)
+				q[i] = s
+			}
 		}
-		time.Sleep(sw.o.MemoryInterval)
 	}
+	return query.Encode()
 }
 
-// ServeHTTP is a middleware for sending meta information to segment.
-func (sw *Service) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
-	var start time.Time
-	if !sw.optOut {
-		start = time.Now()
-	}
-
-	next(rw, r)
-
-	if sw.optOut {
-		return
-	}
-
-	latency := time.Now().UTC().Sub(start.UTC()) / time.Millisecond
-
+func enqueue(
```

Comment:
Yeah! Originally I decoupled it because I used the middleware separately / reused it. Since I've changed my approach, this stayed the same. I'll change it back to how it was.